### DISCUSSION TITLE
Requirement records for Rspress v2.0.0

### DISCUSSION THREAD

**Discussion by Timeless0911** [web-infra-dev/rspress#1105] [Announcements]
This thread is used to collect planned breaking changes and notable changes for Rspress v2.0.0 in the future

**Comment by Timeless0911** [comment]
## Upgrade Rspack / Rsbuild v1.0.0

Rspress v2.0.0 will upgrade the dependent `@rsbuild/core` to v1.0.0 which will upgrade the dependent `@rspack/core` to v1.0.0.

Rspack v1.0.0 contains some breaking changes, see https://github.com/web-infra-dev/rspack/discussions/6315 for more details.

**Reply by Timeless0911** [reply]
`@rsbuild/plugin-less` and `@rsbuild/plugin-sass` now are built-in in rspress to avoid breaking changes. May be we can remove them in 2.0.

**Comment by Timeless0911** [comment]
## Upgrade React 19

Support react 19 and maintain compatibility with react 18.

**Comment by Timeless0911** [comment]
## Upgrade md and mdx related dependencies

Below dependencies may be need to upgrade:

- @mdx-js/loader
- @mdx-js/mdx
- @mdx-js/react
- hast-util-from-html
- mdast-util-mdxjs-esm
- rehype-external-links
- rehype-stringify"
- remark
- remark-gfm
- remark-html
- remark-parse
- remark-rehype
- unified
- unist-util-visit
- unist-util-visit-children
- @types/mdast

**Reply by chenjiahan** [reply]
Rspress uses mdx-rs by default, can we make the mdx-js packages as peer dependencies of an optional Rspress plugin? This can greatly reduce the dependencies of Rspress core.

**Reply by Timeless0911** [reply]
Now we still need mdx-js's AST parse to do something now, and `@mdx-js/react` is needed which provide `MDXProvider` in runtime.

We should sort out the related capabilities provided by mdx-js and port same features in rust mode.

**Reply by chenjiahan** [reply]
How about the remark-related packages? can we remove them

**Reply by Timeless0911** [reply]
If so, users who enable `mdxRs: false` need to manually install many deps in `package.json`?

**Comment by Timeless0911** [comment]
## Reusable theme scalability support and theme inheritance

- #901

**Reply by Timeless0911** [reply]
- https://github.com/web-infra-dev/rspress/issues/1215

**Reply by Timeless0911** [reply]
- https://github.com/web-infra-dev/rspress/issues/1263

**Reply by Timeless0911** [reply]
components of built-in theme export 
- https://github.com/web-infra-dev/rspress/discussions/1388

**Reply by Timeless0911** [reply]
custom layout with pageType  https://vitepress.dev/reference/default-theme-layout#custom-layout

**Comment by Timeless0911** [comment]
## Better AST analyze when flatten mdx content

see https://github.com/web-infra-dev/rspress/issues/1071#issuecomment-2106709717

**Comment by Timeless0911** [comment]
## Search Enhancement

- code block and mdx components content search #487 #936 #1065 #1543
- search weight control #942 
- stable remote search API
- better distinguish the difference between remote and local search in rendering logic to avoid slow remote search causing long overall loading time of local search

**Comment by Timeless0911** [comment]
## Refactor for playground and preview plugin

Maybe we can unified them or provide a better experience when used together

**Reply by Timeless0911** [reply]
- #1113 
- https://github.com/web-infra-dev/rspress/issues/1315

**Reply by Timeless0911** [reply]
![img_v3_02bg_d3dd3739-36a5-4d90-b644-328e03830ccg](https://github.com/web-infra-dev/rspress/assets/50201324/ae9d8b1e-83ee-47d2-b40d-f4df30072845)
 better preview UI

**Reply by Timeless0911** [reply]
port to mdx-rs mode

**Reply by SoonIter** [reply]
- https://github.com/web-infra-dev/rspress/issues/1394
iframe for playground

**Comment by Timeless0911** [comment]
## Support generate sitemap file in build phase

It can be implemented by a plugin.

**Reply by chenjiahan** [reply]
Try this: https://github.com/jl917/rspress-plugin-sitemap

**Comment by SoonIter** [comment]
## Support _meta.{js,ts}

- https://github.com/web-infra-dev/rspress/pull/1044

**Comment by Timeless0911** [comment]
## Use multicompiler to generate client and server build

- https://github.com/web-infra-dev/rspress/issues/1167

**Comment by Timeless0911** [comment]
## Update flexsearch version and optimize the search UX

**Comment by Timeless0911** [comment]
## Lazy compilation by default

**Comment by Timeless0911** [comment]
## Fix some mdx-rs issues

https://github.com/web-infra-dev/mdx-rs/issues

**Comment by Timeless0911** [comment]
## Add / Optimize more built-in components

- https://github.com/web-infra-dev/rspress/discussions/1218
- Display in column
- TOC collection https://github.com/web-infra-dev/rspress/discussions/1460

**Comment by chenjiahan** [comment]
## Add Tailwind CSS Prefix

If Rspress users want to use tailwindcss, the built-in tailwindcss of Rspress will conflict with the user's tailwindcss styles. We can add tailwindcss prefix to avoid the conflict.

https://tailwindcss.com/docs/configuration#prefix

**Reply by Timeless0911** [reply]
- #1269

**Comment by SoonIter** [comment]
## apiParseTool

apiParseTool of `@rspress/plugin-api-docgen` uses `react-docgen` rather than `react-docgen-typescript`
- https://github.com/web-infra-dev/rspress/issues/1290

**Comment by chenjiahan** [comment]
## Strict SSG

Make SSG strict by default, remove the `ssg.strict` config and add `ssg.fallback` config.

see https://github.com/web-infra-dev/rspress/issues/1317 for more details.

**Comment by Timeless0911** [comment]
## Container Optimize

Make `details` expandable to enlarge the unimportant content of the paragraph.

**Reply by Timeless0911** [reply]
- https://github.com/web-infra-dev/rspress/issues/1513

**Comment by chenjiahan** [comment]
## Upgrade React Router 7

https://remix.run/blog/react-router-v7

**Comment by SoonIter** [comment]
## Searchable code block

https://github.com/web-infra-dev/rspress/issues/1543

**Comment by markH5** [comment]
when @rspress/plugin-shiki is upgraded to v2.0, will upgrade shiki to the new version?

**Reply by Timeless0911** [reply]
Shiki will be the default highlighting solution in v2.0 and will be upgraded to v3.

**Reply by markH5** [reply]
Thanks

**Comment by SoonIter** [comment]
## Unify all the _runtime_ translation text in `i18n.json`

1. theme-default text key 

<img width="400" alt="image" src="https://github.com/user-attachments/assets/dcb54671-bf7d-4d16-8c4f-b52f1d9698a9" />

2. plugin-algolia text key

3. other runtime components should can be registered in `i18n.json`

and it can be consumed by

```ts
const { t } = useI18n();
```
